Sources/EmbraceCore/Capture/Network/NetworkPayloadCapture/URLSessionTaskCaptureRule.swift

```diff
@@ -0,0 +1,106 @@
+//
+//  Copyright © 2024 Embrace Mobile, Inc. All rights reserved.
+//
+
+import Foundation
+import EmbraceConfigInternal
+
+class URLSessionTaskCaptureRule {
+
+    private var rule: NetworkPayloadCaptureRule
+    private let regex: NSRegularExpression?
+    let publicKey: String
+
+    var id: String {
+        return rule.id
+    }
+    var urlRegex: String {
+        return rule.urlRegex
+    }
+    var expirationDate: Date {
+        return rule.expirationDate
+    }
+
+    init(rule: NetworkPayloadCaptureRule) {
+        self.rule = rule
+
+        do {
+            regex = try NSRegularExpression(pattern: rule.urlRegex, options: .caseInsensitive)
+        } catch {
+            Embrace.logger.error("Error trying to create regex \"\(rule.urlRegex)\" for rule \(rule.id)!\n\(error.localizedDescription)")
+            regex = nil
+        }
+
+        self.publicKey = URLSessionTaskCaptureRule.sanitize(rule.publicKey)
+    }
+
+    func shouldTriggerFor(request: URLRequest?, response: URLResponse?, error: Error?) -> Bool {
+
+        guard let request = request,
+              let url = request.url,
+              let method = request.httpMethod,
+              expirationDate > Date() else {
+            return false
+        }
+
+        // ignore requests to Embrace's back end
+        if let endpoints = Embrace.client?.options.endpoints {
+            guard !url.absoluteString.contains(endpoints.baseURL) else {
+                return false
+            }
+        }
+
+        // check that the url matches
+        guard urlMatches(url.absoluteString) else {
+            return false
+        }
+
+        // check that the method matches
+        if let methods = rule.methods {
+            guard methods.contains(method) else {
+                return false
+            }
+        }
+
+        // check status codes
+        if let statusCodes = rule.statusCodes {
+
+            // if -1 is passed as a status code for the rule
+            // it means we should capture any requests that errors
+            if statusCodes.contains(-1) && error != nil {
```

Comment:
Is the `error` here what URLSession gives back? If so, it's likely that there is nothing to capture. These errors mean that no connection to the server was made, or the network request never even left the device. There isn't much an application developer can do when these errors are received, other than including retry logic.

For instance, if a `404` status code is returned there would not be an `error`. Is it intended for a value of `-1` be used to capture this situation?

Reply:
This was a definition from product:
"-1 is used to capture errors that prevented the requests from leaving the device"